### ISSUE TITLE
Add support for Correlation IDs

### ISSUE THREAD

**Issue by jan-martens** [containous/traefik#4640]
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?

A few weeks ago, i created PR #4483 which adds support for Correlation IDs to Traefik which was closed. @juliens suggested that this might be worth discussing in a separate issue again, so here it is. 

To recap my reasoning for this feature request:

> In our environment we make heavy use of Correlation IDs to track requests throughout our mircoservice infrastructure. We recently moved from Nginx to Traefik and it brought us a lot of benefits, but the possibility to add custom headers is somewhat limited compared to Nginx. We need to add a new random UUID to every requests that comes in from outside our infrastructure to properly track it's path through the environment. This could have been done with the help of custom request headers, but those have some disadvantages:
> 
> If the header is already present, it will get overwritten. Also you cannot generate a new random UUID for every request. All you could do is generate a random UUID for every frontend each time Traefik reloads the configuration ( with the help of the spring template library ) but then you can have a lot of requests with the same correlation id. This is why we wrote a [http middleware](https://gitlab.com/JanMa/correlation) that adds these headers to every incoming request and would like to add it here.

There has already been PR #2290 that addressed a similar issue but it was closed at the time in favor of supporting OpenTracing. Seeing that support for OpenTracing is now present in Traefik I would like to re-open the discussion of adding Correlation ID support to Traefik. 

> In my opinion the use case for a correlation id differs quite a lot from using a tracing engine. With correlation ids it is very easy to track an external request through all service calls in your environment. You "just" have to make sure your applications log the corresponding header name in their access logs and can easily gather all relevant log entries in a tool like Kibana. Having this information in your logging stack is incredibly useful for debugging, especially in a microservice environment. Using a tracing engine on the other hand is useful when you want to gather low level insights of your inter-service communication and see for example how long single remote procedure call takes. Also implementing full tracing support in all your services takes a lot more effort.

I think having both features available in Traefik would be very beneficial. Please let me know what your thoughts are about this :-)

Kind regards,
Jan

**Comment by jan-martens** [comment]
Any ideas/feedback for my proposal?

Kind regards,
Jan

**Comment by Sarke** [comment]
@JanMa, here's a workaround I've shared this previously that you can use.

You can enable the Jaeger tracer and set the sampling rate to 0, but it will still send the header.  Some quick benchmarks show that it doesn't add any significant overhead.

```toml
[tracing]
  [tracing.jaeger]
    samplingParam = 0
    traceContextHeaderName = "X-Request-ID"
```

or if you use enviroment variables like me:

```yml
TRAEFIK_TRACING: "true"
TRAEFIK_TRACING_JAEGER_SAMPLINGPARAM: 0
TRAEFIK_TRACING_JAEGER_TRACECONTEXTHEADERNAME: "X-Request-ID"
```

**Comment by danmichaelo** [comment]
Thanks for sharing that workaround, @Sarke ! It doesn't seem like it's possible to add the ID to Traefik's access logs though? Which is a pity, since my plan was to correlate access log entries from Traefik with logs from the services behind it :/

**Comment by hed854** [comment]
The mentioned workaround indeed works, but Jaeger / Open Tracing traceIDs are not UUIDv4 compliant (see: https://www.jaegertracing.io/docs/next-release/faq/#why-do-jaeger-trace-ids-look-differently-in-kafka-and-in-the-ui)

Do you plan to add support for UUIDv4 correlation ids like @JanMa suggested? It would be beneficial for brownfield architectures where many internal microservices / libraries / features using UUIDv4.

**Comment by rtribotte** [comment]
Hello,

> Any update towards correlation ids (with opentelemetry) for traefik v3 ?

This issue looks to be related to https://github.com/traefik/traefik/issues/8599 (access logs) and https://github.com/traefik/traefik/issues/10221 (logs).
We are planning to support OTEL logs (https://github.com/traefik/traefik/issues/10056) and the OTEL log data model includes the trace and span IDs (https://github.com/open-telemetry/opentelemetry-specification/blob/v1.31.0/specification/logs/data-model.md#field-traceid), and we think that we could leverage this to correlate logs with traces.

But if this issue goes beyond trace propagation, how would that work?
I'm struggling to find documentation/a standard for a such correlation ID (or request ID).
If anyone has some knowledge about this and could explain it, it would be gladly welcome! Thanks!

**Comment by ppuschmann** [comment]
@rtribotte the idea of the correlation-id is a follows:

- on every incoming request with an existing `correlation-id` header, the content is logged and the header is passed to backend-services
- on every incoming request that has no `correlation-id` header, the webserver/proxy would generate one

While debugging you can then search for the unique correlation-id in your logging tool and will then get the complete request chain of your various proxies and microservices.

That was the initial idea behind this proposal.

**Comment by rtribotte** [comment]
Hello @ppuschmann,

Thanks for the explanations!
Today, a few [plugins](https://plugins.traefik.io/plugins) are handling this.

This issue remains open, we are still eager to gather feedback on this, especially for use cases without tracing.

**Comment by schewara** [comment]
@rtribotte 

>  I'm struggling to find documentation/a standard for a such correlation ID (or request ID).

For me, the best solution in scenarios where there is none of the standard tracing solutions in place, would be to use the [W3C Trace Context](https://www.w3.org/TR/trace-context-1/) specification for interoperability.

Based on the individual environments, the services can then pick up and log the unmodified headers (`traceparent` or `tracestate`) for correlation.

OpenTelemetry also has a list of official [Propagators](https://opentelemetry.io/docs/specs/otel/context/api-propagators/#propagators-distribution) they support in addition to the W3C one.

**Comment by Lp-Francois** [comment]
Hello! I am also looking at a Trace ID header being generated if not present on the request :)

It would help to get e2e visibility, especially from the frontend, as we could just check this ID to understand where the problem comes from.

**Comment by nmengin** [comment]
Hello there,

Thank you for the feedback, we have set the status to enhancement.

However, this would not make it to our roadmap as we are focused elsewhere. If a community member would like to build it, let us know, and we will work with you to ensure you have all the information needed to merge it.

Please reach out to us before starting working on it.
We prefer to work with our community members at the beginning of the design process to ensure that we are aligned and can move quickly with the review and merge process. Let us know here before you start, and we will work with you there.

Don’t forget to check out the [contributor docs](https://github.com/traefik/contributors-guide/blob/master/pr_guidelines.md) and link the PR to this issue.

**Comment by V-Legendre** [comment]
Hello there,
Big +1 on this issue 🙂 

As anyone found a workaround ? Maybe a plugin to recommend ?
